src/util.js

```diff
@@ -2,43 +2,47 @@
 const EMPTY = {};
 
 export function exec(url, route, opts=EMPTY) {
-	let reg = /(?:\?([^#]*))?(#.*)?$/,
-		c = url.match(reg),
+	let reg = /^([^?]*)(?:\?([^#]*))?(#.*)?$/,
+		[, pathname, search] = (url.match(reg) || []),
 		matches = {},
 		ret;
-	if (c && c[1]) {
-		let p = c[1].split('&');
-		for (let i=0; i<p.length; i++) {
-			let r = p[i].split('=');
-			matches[decodeURIComponent(r[0])] = decodeURIComponent(r.slice(1).join('='));
-		}
-	}
-	url = segmentize(url.replace(reg, ''));
+	url = segmentize(pathname);
 	route = segmentize(route || '');
 	let max = Math.max(url.length, route.length);
 	for (let i=0; i<max; i++) {
 		if (route[i] && route[i].charAt(0)===':') {
-			let param = route[i].replace(/(^\:|[+*?]+$)/g, ''),
-				flags = (route[i].match(/[+*?]+$/) || EMPTY)[0] || '',
+			let [, param, flags] = /^:(.*?)([+*?]*)$/.exec(route[i]),
 				plus = ~flags.indexOf('+'),
 				star = ~flags.indexOf('*'),
 				val = url[i] || '';
 			if (!val && !star && (flags.indexOf('?')<0 || plus)) {
 				ret = false;
 				break;
 			}
-			matches[param] = decodeURIComponent(val);
-			if (plus || star) {
-				matches[param] = url.slice(i).map(decodeURIComponent).join('/');
+			else if (plus || star) {
+				matches[param] = decodeURIComponent(url.slice(i).join('/'));
 				break;
 			}
+			matches[param] = decodeURIComponent(val);
 		}
 		else if (route[i]!==url[i]) {
 			ret = false;
 			break;
 		}
 	}
 	if (opts.default!==true && ret===false) return false;
+
+	if (search) {
+		const queryParams = {};
+		search.split('&').forEach(parameter => {
+			let [name, ...value] = parameter.split('=');
+			queryParams[decodeURIComponent(name)] = decodeURIComponent(value.join('='));
```

Comment:
I had meant instead of merging `queryParams` into `matches`, do `matches.$query = queryParams` to separate the two.  Not sure which is better.  Agreed about pulling out the query param parsing, exec should never need to look at the querystring anyway, so that seems like a good call.  I'm impartial to whether we call `segmentize()` before `exec()` or within it - probably nice to do the split once and re-use, but only if its not annoying.